### PR TITLE
Optimise display methods

### PR DESCRIPTION
**Pull request by zhuxiujia**
impl for #16
I think the time is also important to the performance of the Format string, such as crates the [httpdate](https://crates.io/crates/httpdate), It directly affects HTTP Web framework performance

* bench code impl
```
#[bench]
fn bench_format_date(bench: &mut Bencher) {
    let date = Date{
        year: 2022,
        month: 7,
        day: 10
    };
    bench.iter(|| {
        date.to_string();
    })
}

#[bench]
fn bench_format_time(bench: &mut Bencher) {
    let time = Time{
        hour: 10,
        minute: 11,
        second: 12,
        microsecond: 11
    };
    bench.iter(|| {
        time.to_string();
    })
}

#[bench]
fn bench_format_date_time(bench: &mut Bencher) {
    let date = DateTime{
        date: Date {
            year: 2022,
            month: 7,
            day: 10
        },
        time: Time {
            hour: 0,
            minute: 0,
            second: 0,
            microsecond: 0
        },
        offset: Some(60)
    };
    bench.iter(|| {
        date.to_string();
    })
}
```

* befor  change
```
test bench_format_date               ... bench:         149 ns/iter (+/- 3)
test bench_format_time               ... bench:         231 ns/iter (+/- 4)
test bench_format_date_time          ... bench:         283 ns/iter (+/- 3) //offset None
test bench_format_date_time          ... bench:         326 ns/iter (+/- 6) //offset Some(60)
```
* after change
```
test bench_format_date               ... bench:          51 ns/iter (+/- 0)
test bench_format_time               ... bench:          53 ns/iter (+/- 3)
test bench_format_date_time          ... bench:          50 ns/iter (+/- 0) //offset None
test bench_format_date_time          ... bench:         115 ns/iter (+/- 0) //offset Some(60)
```

The DateTime format will affect about 50ns due to offset, because it is divided twice of  Memory allocation
```
f.write_str(std::str::from_utf8(&buf[..]).unwrap())?;
```
so.If I can fix the DateTime  format string length to Fixed length, then this performance can be improved to 50ns/iter

### CODE REVIEW

File: benches/main.rs [hunk 1]
```diff
@@ -201,3 +201,50 @@ fn x_combined(bench: &mut Bencher) {
...
+
+#[bench]
+fn bench_format_date(bench: &mut Bencher) {
+    let date = Date{
+        year: 2022,
+        month: 7,
+        day: 10
+    };
+    bench.iter(|| {
+        date.to_string();
```
Comment (samuelcolvin) [inline benches/main.rs:214]: You been to use black_box here, probably for both the input and the result.

Comment (zhuxiujia) [reply benches/main.rs]: Yes, I have added this code

File: benches/main.rs [hunk 2]
```diff
@@ -201,3 +201,49 @@ fn x_combined(bench: &mut Bencher) {
         black_box(Time::parse_str(&t4).unwrap());
     })
 }
+
+#[bench]
+fn bench_format_date(bench: &mut Bencher) {
```
Comment (samuelcolvin) [inline benches/main.rs:206]:
```suggestion
fn format_date(bench: &mut Bencher) {
```

Bit pretty of me, but it's somewhat tautologous to say "bench_" we we know these are benchmarks.

File: benches/main.rs [hunk 3]
```diff
@@ -201,3 +201,49 @@ fn x_combined(bench: &mut Bencher) {
...
+        month: 7,
+        day: 10,
+    });
+    bench.iter(|| {
+        black_box(date.to_string());
+    })
+}
+
+#[bench]
+fn bench_format_time(bench: &mut Bencher) {
```
Comment (samuelcolvin) [inline benches/main.rs:218]:
```suggestion
fn format_time(bench: &mut Bencher) {
```

File: benches/main.rs [hunk 4]
```diff
@@ -201,3 +201,49 @@ fn x_combined(bench: &mut Bencher) {
...
+        second: 12,
+        microsecond: 11,
+    });
+    bench.iter(|| {
+        black_box(time.to_string());
+    })
+}
+
+#[bench]
+fn bench_format_date_time(bench: &mut Bencher) {
```
Comment (samuelcolvin) [inline benches/main.rs:231]:
```suggestion
fn format_date_time(bench: &mut Bencher) {
```

File: src/datetime.rs [hunk 5]
```diff
@@ -33,13 +32,70 @@ pub struct DateTime {
...
+            buf[6] = b'0' + (self.date.month % 10) as u8;
+            buf[8] = b'0' + (self.date.day / 10) as u8;
+            buf[9] = b'0' + (self.date.day % 10) as u8;
+            buf[11] = b'0' + (self.time.hour / 10) as u8;
+            buf[12] = b'0' + (self.time.hour % 10) as u8;
+            buf[14] = b'0' + (self.time.minute / 10) as u8;
+            buf[15] = b'0' + (self.time.minute % 10) as u8;
+            buf[17] = b'0' + (self.time.second / 10) as u8;
+            buf[18] = b'0' + (self.time.second % 10) as u8;
+            buf[19] = b'.';
```
Comment (samuelcolvin) [inline src/datetime.rs:53]: you can surely remove this?

File: src/time.rs [hunk 6]
```diff
@@ -38,12 +38,38 @@ pub struct Time {
...
-            write!(f, ".{}", s.trim_end_matches('0'))?;
+            let mut buf: [u8; 15] = *b"00:00:00.000000";
+            buf[0] = b'0' + (self.hour / 10) as u8;
+            buf[1] = b'0' + (self.hour % 10) as u8;
+            buf[3] = b'0' + (self.minute / 10) as u8;
+            buf[4] = b'0' + (self.minute % 10) as u8;
+            buf[6] = b'0' + (self.second / 10) as u8;
+            buf[7] = b'0' + (self.second % 10) as u8;
+
+            buf[8] = b'.';
```
Comment (samuelcolvin) [inline src/time.rs:55]: as above.

File: src/time.rs [hunk 7]
```diff
@@ -38,12 +38,38 @@ pub struct Time {
 
 impl fmt::Display for Time {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        write!(f, "{:02}:{:02}:{:02}", self.hour, self.minute, self.second)?;
+        // write!(f, "{:02}:{:02}:{:02}", self.hour, self.minute, self.second)?;
```
Comment (samuelcolvin) [inline src/time.rs:41]: please remove.

**Review by samuelcolvin — COMMENTED** [review]
humm, is there anyway to reduce the amount of boiler plate, e.g. with macros or with a loop?

Currently there's a lot of repetition.

File: src/datetime.rs (hunk 5 above)
Comment (zhuxiujia) [reply src/datetime.rs]: I can cut out some boilerplate code.But complete removal affects performance，Because buffer is pre-allocated

**Comment by samuelcolvin** [comment]
thanks so much 🙏 🚀 .

**Pull request merged**